From 0e913f9df1e34e8200f43e3746f3f4abd1549631 Mon Sep 17 00:00:00 2001
From: yuyue <yuyue@mprc.pku.edu.cn>
Date: Wed, 7 Mar 2012 11:19:31 +0800
Subject: [PATCH 319/641] UniCore64: change the put function

We use two general registers to store r0 and r1 for saving the
current state, but now we replace r0 and r1 with p0.c12 which are
backup registers.

Signed-off-by: yuyue <yuyue@mprc.pku.edu.cn>
---
 arch/unicore64/include/arch/asm-debug.h |   24 ++++++++++--------------
 arch/unicore64/kernel/entry.S           |    4 ++--
 arch/unicore64/lib/debug.S              |    6 +++---
 arch/unicore64/lib/mcount.S             |   10 ++++------
 4 files changed, 19 insertions(+), 25 deletions(-)

diff --git a/arch/unicore64/include/arch/asm-debug.h b/arch/unicore64/include/arch/asm-debug.h
index b84f76a..25f9fb7 100644
--- a/arch/unicore64/include/arch/asm-debug.h
+++ b/arch/unicore64/include/arch/asm-debug.h
@@ -9,24 +9,20 @@
  */
 #ifdef	CONFIG_DEBUG_OCD
 /* For OCD debug mode */
-__ASMMACRO_WRAP(.macro	__putchar, rchar, rt1, rt2;
-			mov	&rt1, r0;
-			mov	&rt2, r1;
-			mov	r1, &rchar;
-			mov	r0, #1;
+__ASMMACRO_WRAP(.macro	__putchar, rchar;
+			movc	p0.c12, &rchar, #7;
+			dmov	&rchar, #1;
+			movc	p0.c12, &rchar, #6;
 			bkpt;
-			mov	r0, &rt1;
-			mov	r1, &rt2;
+			movc	&rchar, p0.c12, #7;
 		.endm)
 
-__ASMMACRO_WRAP(.macro	__putdata, rdata, rt1, rt2;
-			dmov	&rt1, r0;
-			dmov	&rt2, r1;
-			dmov	r1, &rdata;
-			dmov	r0, #0;
+__ASMMACRO_WRAP(.macro	__putdata, rdata;
+			movc	p0.c12, &rdata, #7;
+			dmov	&rdata, #0;
+			movc	p0.c12, &rdata, #6;
 			bkpt;
-			dmov	r0, &rt1;
-			dmov	r1, &rt2;
+			movc	&rdata, p0.c12, #7;
 		.endm)
 #endif /* CONFIG_DEBUG_OCD */
 
diff --git a/arch/unicore64/kernel/entry.S b/arch/unicore64/kernel/entry.S
index 86296b5..7e8d32a 100644
--- a/arch/unicore64/kernel/entry.S
+++ b/arch/unicore64/kernel/entry.S
@@ -41,8 +41,8 @@ ENDPROC(__switch_to)
 
 ENTRY(__vec_invalid)
 	dmovl		r0,  0xdeaddeaddeaddead
-	__putdata	r0, r14, r15
-	__putdata	lr, r14, r15
+	__putdata	r0
+	__putdata	lr
 	__halt		/* no return */
 ENDPROC(__vec_invalid)
 
diff --git a/arch/unicore64/lib/debug.S b/arch/unicore64/lib/debug.S
index 9021f5f..b65e5d7 100644
--- a/arch/unicore64/lib/debug.S
+++ b/arch/unicore64/lib/debug.S
@@ -16,7 +16,7 @@ ENTRY(uc64_debug_puts)
 	dcmpsub.a	r2, #0
 	beq	1003f
 
-	__putchar	r2, r14, r15
+	__putchar	r2
 
 	dsub.a	r1, r1, #1
 	bsl	1002f
@@ -36,11 +36,11 @@ ENDPROC(uc64_debug_puts)
  * r0: the data to be printed
  */
 ENTRY(uc64_debug_putx)
-	__putdata	r0, r14, r15
+	__putdata	r0
 
 	/* ____debug_putx: we add a blank after data */
 	mov	r0, #0x20
-	__putchar	r0, r14, r15
+	__putchar	r0
 
 	return
 ENDPROC(uc64_debug_putx)
diff --git a/arch/unicore64/lib/mcount.S b/arch/unicore64/lib/mcount.S
index 95b0198..bc99e69 100644
--- a/arch/unicore64/lib/mcount.S
+++ b/arch/unicore64/lib/mcount.S
@@ -1,14 +1,12 @@
 #include <linux/linkage.h>
 #include <arch/asm-common.h>
+#include <arch/asm-debug.h>
 
 ENTRY(mcount)
 	__push	r0
-	__push	r1
-	ldd     r1, [ip-], #8
-        dsub    r1, r1, #4
-	mov	r0, #0
-	bkpt
-	__pop	r1
+	ldd     r0, [ip-], #8
+        dsub    r0, r0, #4
+	__putdata	r0
 	__pop	r0
 	return
 ENDPROC(mcount)
-- 
1.7.9.5

